### PR TITLE
WIP: feat: karma system

### PR DESCRIPTION
**Pull request by rickklaasboer**
image
image
image
image

### CODE REVIEW

File: src/commands/karma/KarmaUserResetCommand.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    TextInputStyle,
+} from 'discord.js';
+
+export const KarmaUserResetCommand = new Command({
+    name: 'internal',
+    description: 'internal',
+    run: async (interaction) => {
+        const user = interaction.options.getUser('user') ?? interaction.user;
+
+        const isAdmin = interaction.memberPermissions?.has(
```
Comment (Jessuhh) [inline src/commands/karma/KarmaUserResetCommand.ts:18]: Maybe also only allow admins to enable and disable karma channels? ;)

File: src/commands/karma/KarmaUserResetCommand.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        modalBuilder.addComponents(actionRowBuilder);
+
+        await interaction.showModal(modalBuilder);
+
+        const submitted = await interaction.awaitModalSubmit({
+            time: 60000,
+            filter: (i) => i.user.id === interaction.user.id,
+        });
+
+        if (submitted) {
```
Comment (Kabelbaan) [inline src/commands/karma/KarmaUserResetCommand.ts:57]: reduce nesting?

**Review by Kabelbaan — CHANGES_REQUESTED** [review]

File: src/lang/en.json [hunk 3]
```diff
@@ -1,3 +1,7 @@
 {
-    "errors.common.failed": "An unexpected error occurred while trying to generate your %s, please try again later!"
-}
+    "errors.common.failed": "An unexpected error occurred while trying to generate your %s, please try again later!",
+    "errors.common.command.no_permission": "I'm sorry <@%s>, you don't have permission to execute this command.",
+    "karma.channel.disable.success": "Done! Channel `#%s` is no longer a karma channel.",
+    "karma.channel.enable.success": "Done! Channel `#%s` is now a karma channel.",
+    "karka.channel.reset.not_a_karma_channel": "Whoops, looks like `#%s` is not a karma channel."
```
Comment (Jessuhh) [inline src/lang/en.json:6]: `karka` => `karma`

**Review by Jessuhh — CHANGES_REQUESTED** [review]

File: src/commands/karma/KarmaChannelDisableCommand.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+export const KarmaChannelDisableCommand = new Command({
+    name: 'internal',
+    description: 'internal',
+    run: async (interaction) => {
+        const {id, name = ''} = interaction.options.getChannel('channel') ?? {};
+
+        if (!isAdmin(interaction)) {
+            await interaction.reply(
+                t('errors.common.command.no_permission', interaction.user.id),
```
Comment (NiekKleverkamp) [inline src/commands/karma/KarmaChannelDisableCommand.ts:15]: Make ephemeral?

File: src/commands/karma/KarmaChannelEnableCommand.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+export const KarmaChannelEnableCommand = new Command({
+    name: 'internal',
+    description: 'internal',
+    run: async (interaction) => {
+        const {id, name = ''} = interaction.options.getChannel('channel') ?? {};
+
+        if (!isAdmin(interaction)) {
+            await interaction.reply(
+                t('errors.common.command.no_permission', interaction.user.id),
```
Comment (NiekKleverkamp) [inline src/commands/karma/KarmaChannelEnableCommand.ts:15]: Make ephemeral?

File: src/commands/karma/KarmaUserResetCommand.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+export const KarmaUserResetCommand = new Command({
+    name: 'internal',
+    description: 'internal',
+    run: async (interaction) => {
+        const user = interaction.options.getUser('user') ?? interaction.user;
+
+        if (!isAdmin(interaction)) {
+            await interaction.reply(
+                t('errors.common.command.no_permission', interaction.user.id),
```
Comment (NiekKleverkamp) [inline src/commands/karma/KarmaUserResetCommand.ts:44]: Make ephemeral?

File: src/commands/karma/KarmaUserResetCommand.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            time: 60000,
+            filter: (i) => i.user.id === interaction.user.id,
+        });
+
+        if (!submitted) return;
+
+        const username = submitted.fields.getTextInputValue('USERNAME_INPUT');
+
+        if (username !== user.username) {
+            await submitted.reply('Usernames did not match, please try again.');
```
Comment (NiekKleverkamp) [inline src/commands/karma/KarmaUserResetCommand.ts:62]: Make ephemeral?

File: src/commands/karma/KarmaUserResetCommand.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            return;
+        }
+
+        const rowsAffected = await Karma.query()
+            .where('guildId', '=', interaction.guildId)
+            .andWhere('userId', '=', user.id)
+            .delete();
+
+        await submitted.reply(
```
Comment (NiekKleverkamp) [inline src/commands/karma/KarmaUserResetCommand.ts:71]: Could be made ephemeral I suppose :P

**Review by NiekKleverkamp — CHANGES_REQUESTED** [review]
Ephemeral errors pl0x

**Review by NiekKleverkamp — APPROVED** [review]
lgtm

**Review by Jessuhh — APPROVED** [review]
LGTM! Is there a reason why the enable and disable karma channel command are separated? Could you not make one command do both?

**Pull request merged**